File: prime/src/test/java/com/telenordigital/prime/disruptor/PrimeEventProducerTest.java
package com.telenordigital.prime.disruptor;

import com.lmax.disruptor.EventHandler;
import com.lmax.disruptor.RingBuffer;
import com.lmax.disruptor.dsl.Disruptor;
import com.telenordigital.prime.ocs.CreditControlRequestInfo;
import com.telenordigital.prime.ocs.MultipleServiceCreditControl;
import com.telenordigital.prime.ocs.ReguestedServiceUnit;
import com.telenordigital.prime.ocs.UsedServiceUnit;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import static com.telenordigital.prime.disruptor.PrimeEventMessageType.CREDIT_CONTROL_REQUEST;
import static com.telenordigital.prime.disruptor.PrimeEventMessageType.TOPUP_DATA_BUNDLE_BALANCE;
import static junit.framework.Assert.assertFalse;
import static junit.framework.TestCase.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class PrimeEventProducerTest {

    private static final long NO_OF_TOPUP_BYTES = 991234L;

    private static final String MSISDN = "+4711223344";

    private static final String STREAM_ID = "mySecret stream";

    private static final int RING_BUFFER_SIZE = 256;

    private static final int TIMEOUT = 10;

    private  PrimeEventProducer pep;

    private Disruptor<PrimeEvent> disruptor;

    private CountDownLatch cdl;

    private Set<PrimeEvent> result;


    @SuppressWarnings("unchecked")
    @Before
    public void setUp() {
        this.disruptor = new Disruptor<PrimeEvent>(
            ()-> new PrimeEvent(),
            RING_BUFFER_SIZE,
            Executors.defaultThreadFactory() );
        final RingBuffer<PrimeEvent> ringBuffer = disruptor.getRingBuffer();
        this.pep = new PrimeEventProducer(ringBuffer);

        this.cdl = new CountDownLatch(1);
        this.result = new HashSet<>();
        final EventHandler<PrimeEvent> eh = (event, sequence, endOfBatch) -> {
            result.add(event);
            cdl.countDown();
        };

        //noinspection unchecked
        disruptor.handleEventsWith(eh);
        disruptor.start();
    }

    private PrimeEvent getCollectedEvent()  throws InterruptedException {
        // Wait  wait a short while for the thing to process.
        assertTrue(cdl.await(TIMEOUT, TimeUnit.SECONDS));
        assertFalse(result.isEmpty());
        final PrimeEvent event = result.iterator().next();
        assertNotNull(event);
        return event;
    }

    @After
    public void shutDown() {
        disruptor.shutdown();
    }

    @Test
    public void topupDataBundleBalanceEvent() throws Exception {

        // Stimulating a response
        pep.topupDataBundleBalanceEvent(MSISDN, NO_OF_TOPUP_BYTES);

        // Collect an event (or fail trying).
        final PrimeEvent event = getCollectedEvent();

        // Verify some behavior
        assertEquals(MSISDN, event.getMsisdn());
        assertEquals(NO_OF_TOPUP_BYTES, event.getRequestedBucketBytes());
        assertEquals(TOPUP_DATA_BUNDLE_BALANCE, event.getMessageType());
    }

    @Test
    public void creditControlRequestEvent() throws Exception {
        final CreditControlRequestInfo request =
                CreditControlRequestInfo.
                        newBuilder().
                        setMsisdn(MSISDN).
                        addMscc(MultipleServiceCreditControl.newBuilder()
                                .setRequested(ReguestedServiceUnit.newBuilder()
                                        .setTotalOctets(NO_OF_TOPUP_BYTES)
                                        .build())
                                .setUsed(UsedServiceUnit.newBuilder().setTotalOctets(0L).build())
                                .setRatingGroup(10)
                                .setServiceIdentifier(1)
                                .build()
                        ).build();

        pep.injectCreditControlRequestIntoRingbuffer(request, STREAM_ID);

        final PrimeEvent event = getCollectedEvent();
        assertEquals(MSISDN, event.getMsisdn());
        assertEquals(NO_OF_TOPUP_BYTES, event.getRequestedBucketBytes());
        assertEquals(0L, event.getUsedBucketBytes());
        assertEquals(10, event.getRatingGroup());
        assertEquals(1, event.getServiceIdentifier());
        assertEquals(STREAM_ID, event.getOcsgwStreamId());
        assertEquals(CREDIT_CONTROL_REQUEST, event.getMessageType());
    }
}

File: ocs/src/test/kotlin/org/ostelco/prime/disruptor/PrimeEventProducerTest.kt
package org.ostelco.prime.disruptor

import com.lmax.disruptor.EventFactory
import com.lmax.disruptor.EventHandler
import com.lmax.disruptor.dsl.Disruptor
import junit.framework.TestCase.assertEquals
import org.junit.After
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.ostelco.ocs.api.CreditControlRequestInfo
import org.ostelco.ocs.api.MultipleServiceCreditControl
import org.ostelco.ocs.api.ServiceUnit
import org.ostelco.prime.disruptor.EventMessageType.CREDIT_CONTROL_REQUEST
import org.ostelco.prime.disruptor.EventMessageType.TOPUP_DATA_BUNDLE_BALANCE
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class PrimeEventProducerTest {

    private lateinit var primeEventProducer: EventProducerImpl

    private var disruptor: Disruptor<OcsEvent>? = null

    private lateinit var countDownLatch: CountDownLatch

    private lateinit var result: MutableSet<OcsEvent>

    private// Wait a short while for the thing to process.
    val collectedEvent: OcsEvent
        get() {
            assertTrue(countDownLatch.await(TIMEOUT.toLong(), TimeUnit.SECONDS))
            assertFalse(result.isEmpty())
            val event = result.iterator().next()
            assertNotNull(event)
            return event
        }


    @Before
    fun setUp() {
        this.disruptor = Disruptor(
                EventFactory<OcsEvent> { OcsEvent() },
                RING_BUFFER_SIZE,
                Executors.defaultThreadFactory())
        val ringBuffer = disruptor?.ringBuffer ?: throw Exception("Failed to init disruptor")
        this.primeEventProducer = EventProducerImpl(ringBuffer)

        this.countDownLatch = CountDownLatch(1)
        this.result = HashSet()
        val eh = EventHandler<OcsEvent> { event, _, _ ->
            result.add(event)
            countDownLatch.countDown()
        }

        disruptor?.handleEventsWith(eh)
        disruptor?.start()
    }

    @After
    fun shutDown() {
        disruptor?.shutdown()
    }

    @Test
    fun topupDataBundleBalanceEvent() {

        // Stimulating a response
        primeEventProducer.topupDataBundleBalanceEvent(
                requestId = TOPUP_REQUEST_ID,
                bundleId = BUNDLE_ID,
                bytes = NO_OF_TOPUP_BYTES)

        // Collect an event (or fail trying).
        val event = collectedEvent

        // Verify some behavior
        assertEquals(BUNDLE_ID, event.bundleId)
        assertEquals(NO_OF_TOPUP_BYTES, event.topupContext?.topUpBytes)
        assertEquals(TOPUP_DATA_BUNDLE_BALANCE, event.messageType)
    }

    @Test
    fun creditControlRequestEvent() {
        val request = CreditControlRequestInfo.newBuilder().setMsisdn(MSISDN).addMscc(MultipleServiceCreditControl.newBuilder()
                .setRequested(ServiceUnit.newBuilder()
                        .setTotalOctets(REQUESTED_BYTES)
                        .build())
                .setUsed(ServiceUnit.newBuilder().setTotalOctets(USED_BYTES).build())
                .setRatingGroup(RATING_GROUP)
                .setServiceIdentifier(SERVICE_IDENTIFIER)
                .build()
        ).build()

        primeEventProducer.injectCreditControlRequestIntoRingbuffer(STREAM_ID, request)

        val event = collectedEvent
        assertEquals(MSISDN, event.msisdn)
        assertEquals(REQUESTED_BYTES, event.request?.msccList?.firstOrNull()?.requested?.totalOctets ?: 0L)
        assertEquals(USED_BYTES, event.request?.msccList?.firstOrNull()?.used?.totalOctets ?: 0L)
        assertEquals(RATING_GROUP, event.request?.msccList?.firstOrNull()?.ratingGroup)
        assertEquals(SERVICE_IDENTIFIER, event.request?.msccList?.firstOrNull()?.serviceIdentifier)
        assertEquals(STREAM_ID, event.ocsgwStreamId)
        assertEquals(CREDIT_CONTROL_REQUEST, event.messageType)
    }

    companion object {

        private const val NO_OF_TOPUP_BYTES = 991234L

        private const val REQUESTED_BYTES = 500L

        private const val USED_BYTES = 300L

        private const val BUNDLE_ID = "foo@bar.com"

        private const val MSISDN = "4711223344"

        private const val STREAM_ID = "mySecret stream"

        private const val RING_BUFFER_SIZE = 256

        private const val TIMEOUT = 10

        private const val RATING_GROUP = 10L

        private const val SERVICE_IDENTIFIER = 1L

        private const val TOPUP_REQUEST_ID = "req-id"
    }
}

